include/otf2xx/event/metric.hpp

```diff
@@ -118,121 +120,428 @@ namespace event
         metric(otf2::chrono::time_point timestamp,
-               const otf2::definition::metric_instance& metric_c,
-               const std::vector<value_container>& values)
-        : base<metric>(timestamp), metric_class_(), metric_instance_(metric_c), values_(values)
+               const otf2::definition::metric_instance& metric_c, metric_values&& values)
+        : base<metric>(timestamp), metric_class_(), metric_instance_(metric_c),
+          values_(std::move(values))
+        {
+        }
+
+        metric(OTF2_AttributeList* al, otf2::chrono::time_point timestamp,
+               const otf2::definition::metric_instance& metric_c, metric_values&& values)
+        : base<metric>(al, timestamp), metric_class_(), metric_instance_(metric_c),
+          values_(std::move(values))
+        {
+        }
+
+        // construct without values, but reserve memory for them
+        metric(otf2::chrono::time_point timestamp, const otf2::definition::metric_class& metric_c)
+        : base<metric>(timestamp), metric_class_(metric_c), metric_instance_(), values_(metric_c)
         {
         }
 
         metric(OTF2_AttributeList* al, otf2::chrono::time_point timestamp,
-               const otf2::definition::metric_instance& metric_c,
-               const std::vector<value_container>& values)
-        : base<metric>(al, timestamp), metric_class_(), metric_instance_(metric_c), values_(values)
+               const otf2::definition::metric_class& metric_c)
+        : base<metric>(al, timestamp), metric_class_(metric_c), metric_instance_(),
+          values_(metric_c)
         {
         }
 
         // copy constructor with new timestamp
         metric(const otf2::event::metric& other, otf2::chrono::time_point timestamp)
         : base<metric>(other, timestamp), metric_class_(other.metric_class()),
-          metric_instance_(other.metric_instance()), values_(other.values())
+          metric_instance_(other.metric_instance()), values_(other.raw_values())
         {
         }
 
-        std::vector<value_container>& values()
+        /// construct without referencing a metric class or a metric instance
+        metric(OTF2_AttributeList* al, otf2::chrono::time_point timestamp, metric_values&& values)
+        : base<metric>(al, timestamp), metric_class_(), metric_instance_(),
+          values_(std::move(values))
+        {
+        }
+
+        metric_values& raw_values()
         {
             return values_;
         }
 
-        const std::vector<value_container>& values() const
+        const metric_values& raw_values() const
         {
             return values_;
         }
 
-        const value_container& get_value_for(const otf2::definition::metric_member& member) const
+        value_proxy get_value_at(std::size_t index)
         {
-            std::size_t i = 0;
+            auto metric_class = resolve_metric_class_ref();
 
-            auto mc = metric_class_.lock();
+            assert(metric_class->is_valid());
```

Comment:
These cannot work, did you built in release?

The thing is, a weak_ref cannot check if it has a valid pointer. It can only be checked against nullptr

You'll probably want to check for `metric_class != nullptr` or something akin.